appinventor/blocklyeditor/src/generators/yail/math.js

```diff
@@ -388,17 +388,35 @@ Blockly.Yail['math_format_as_decimal'] = function() {
 };
 
 Blockly.Yail['math_is_a_number'] = function() {
-  // Basic is_a_number.
+  var mode = this.getFieldValue('OP');
+  var tuple = Blockly.Yail.math_is_a_number.OPERATORS[mode];
+  var yailname = tuple[0];
+  var description = tuple[1];
   var argument = Blockly.Yail.valueToCode(this, 'NUM', Blockly.Yail.ORDER_NONE) || Blockly.Yail.YAIL_FALSE;
-  var code = Blockly.Yail.YAIL_CALL_YAIL_PRIMITIVE + "is-number?"
-      + Blockly.Yail.YAIL_SPACER;
-  code = code + Blockly.Yail.YAIL_OPEN_COMBINATION
-      + Blockly.Yail.YAIL_LIST_CONSTRUCTOR + Blockly.Yail.YAIL_SPACER
-      + argument + Blockly.Yail.YAIL_CLOSE_COMBINATION;
-  code = code + Blockly.Yail.YAIL_SPACER + Blockly.Yail.YAIL_QUOTE
-      + Blockly.Yail.YAIL_OPEN_COMBINATION + "any"
-      + Blockly.Yail.YAIL_CLOSE_COMBINATION + Blockly.Yail.YAIL_SPACER;
-  code = code + Blockly.Yail.YAIL_DOUBLE_QUOTE + "is a number?"
-      + Blockly.Yail.YAIL_DOUBLE_QUOTE + Blockly.Yail.YAIL_CLOSE_COMBINATION;
+  var code ="(call-yail-primitive " + yailname + " (*list-for-runtime* " + argument +") " + description + ")";
   return [ code, Blockly.Yail.ORDER_ATOMIC ];
 };
+
+Blockly.Yail.math_is_a_number.OPERATORS = {
+  NUMBER: ['is-number?', "'(any) \"is a number?\""],
+  DECIMAL: ['is-decimal?', "'(any) \"is decimal?\""],
```

Comment:
is-decimal?, is-hex?, is-binary?  all need to have type text rather than type any, e.g.,
HEXADECIMAL: ['is-hexadecimal?', "'(text) \"is hexadecimal?\""],

Observe that, in contrast, is-number? should have type any, as before.  The difference is that the yail primitive is-number can accept any type, while the pattern matcher used by is-hex, etc., will throw a type error if it is not given a string.   So, in your original code, for  example, giving a number as input to the is-hex block signals an error.   By using '(text) here, you will cause app inventor to generate the appropriate type checking code and conversions.   Whenever you create a new primitive, make sure to test it with inputs of all different types, inclucing non-sensible things like lists and even components (e.g., sprites) to make sure the errors are being caught.

Please also add unit tests for the Yail code that checks the patterns and does the conversion.
You can add those in
appinventor/buildserver/tests/com/google/appinventor/buildserver/YailEvalTest
See the testIsNumber as an example -- don't forget to change (any) to (text) as appropriate.